### ISSUE TITLE
[Feature Request] Search mangadex using search terms

### ISSUE THREAD

**Issue by jakoblover** [hkalexling/Mango#42]
**Is your feature request related to a problem? Please describe.**
It is cumbersome to find the url on mangadex if i know exactly the name of the manga i wanna read

**Describe the solution you'd like**
I would like to search mangadex for a manga straight in Mango. If I know exactly what I want, I would like to enter a search term and get a list of candidates. 

**Additional context**
It doesn't have to be very complicated, but a simple search engine would be nice.

**Comment by hkalexling** [comment]
I agree it would be nice feature to have, but as far as I know MangaDex does not have open APIs for searching, so the only way to have the functionality is to parse the HTML directly. This can be problematic and will probably break things when they update their UI :/

@jaredlt What do you think?

**Comment by hkalexling** [comment]
I found a MangaDex JS API that does this by matching manga URLs in the HTML using regex. See https://github.com/md-y/mangadex-full-api/blob/master/src/structure/manga.js#L218-L225

This approach should be more robust. I think it's worth trying out.

**Comment by jaredlt** [comment]
Agree that the API would be best. I just asked in their Discord to check and search still isn't available unfortunately. So looks like scraping is the only option. Some random thoughts and comments:

- How do we ensure the user is logged in?
- Handling errors and rate limits (we may want to explore some form of exponential backoff as we wouldn't want a Mangadex's user account to be banned because another Mango admin keeps trying to search - worst case, but something to consider)
- Rather than regex the entire html would it be better to reach for a html parser and then iterate of each `<a` tag? I see there's a native [.parse_html](https://crystal-lang.org/api/0.22.0/XML.html) method or the [myhtml](https://github.com/kostya/myhtml) shard looks good (maybe overkill for v1)
- The quick search [results page](https://mangadex.org/quick_search/ghost%20in%20the%20shell) is pretty clean html. If we want to bring in eg. Title, Author, Image it may not be too difficult to scrape this data and still do it in a reasonably robust way. 
- I really like the simplicity of the single input form on the Download page and I wonder if we can avoid making it more complex ie. handle all cases (id, url, search term). URL is easy to detect but in it's current form if a manga title is just numbers this will conflict with being able to search. I've seen systems like Radarr get around this by introducing a prefix eg. by default it will search by search term but if you enter `imdb:tt0116481` is will search by id. We could perhaps do something similar eg. `mangadex:5420`

**Comment by jakoblover** [comment]
It would probably not be very hard to have the single search bar. If it's a valid mangadex id and does not return any results specifically for that id, search using search terms.

**Comment by Baitinq** [comment]
I agree that this would be an extremely good addition to mango.

**Comment by hkalexling** [comment]
FYI we had some discussions on improving MangaDex integration at #24. I will work on that soon, and the searching feature will also be implemented as a by-product of #24 :)
